frb_codegen/src/commands.rs

```diff
@@ -1,22 +1,72 @@
 use std::fmt::Write;
 use std::path::Path;
+use std::path::PathBuf;
 use std::process::Command;
 use std::process::Output;
 
 use crate::error::{Error, Result};
+use itertools::Itertools;
 use log::{debug, info, warn};
 
+/// - First argument is either a string of a command, or a function receiving a slice of [`PathBuf`].
+/// - Following arguments are either:
+///   - An expression to turn into a [`PathBuf`]; or
+///   - `?<expr>` to add `expr` only if `expr` is a [`Some`]; or
+///   - A tuple of `(condition, expr, ...expr)` that adds `expr`s to the arguments only if `condition` is satisfied.
+///
+/// Returns [`Output`] if executing a command name, or the return value of the specified function.
+#[macro_export]
+macro_rules! run {
+    ($binary:literal, $($rest:tt)*) => {{
+        let args = $crate::args!($($rest)*);
+        $crate::commands::execute_command($binary, args.iter(), None)
+    }};
+    ($command:path $([ $($args:expr),* ])?, $($rest:tt)*) => {{
+        let args = $crate::args!($($rest)*);
+        $command(&args[..] $(, $($args),* )?)
+    }};
+}
+
+/// Formats a list of [`PathBuf`]s using the syntax detailed in [`run`].
+#[macro_export]
+macro_rules! args {
+    (@args $args:ident $(,)?) => {};
+    (@args $args:ident ($cond:expr, $($expr:expr),+ $(,)?), $($rest:tt)*) => {
+        if $cond {
+            $(
+                $args.push(::std::path::PathBuf::from($expr));
+            )+
+        }
+        $crate::args!(@args $args $($rest)*);
+    };
+    (@args $args:ident ?$src:expr, $($rest:tt)*) => {
+        if let Some(it) = (&$src) {
+            $args.push(::std::path::PathBuf::from(it));
+        }
+        $crate::args!(@args $args $($rest)*);
+    };
+    (@args $args:ident $expr:expr, $($rest:tt)*) => {
+        $args.push(::std::path::PathBuf::from($expr));
+        $crate::args!(@args $args $($rest)*);
+    };
+    ($($rest:tt)*) => {{
+        let mut args = vec![];
+        $crate::args!(@args args $($rest)*,);
+        args
+    }};
+}
+
 #[must_use]
 fn call_shell(cmd: &str) -> Output {
     #[cfg(windows)]
-    return execute_command("powershell", &["-noprofile", "-c", cmd], None);
+    return run!("powershell", "-noprofile", "-c", cmd);
 
     #[cfg(not(windows))]
-    execute_command("sh", &["-c", cmd], None)
+    run!("sh", "-c", cmd)
 }
 
 pub fn ensure_tools_available() -> Result {
-    let output = call_shell("dart pub global list");
+    let output = run!("dart", "pub", "global", "list");
```

Comment:
Looks like it is non-equivalent? b/c originally we call `call_shell`, and thus, we use `powershell -c` or `sh -c`; but now, we directly call it